libsolidity/interface/CompilerStack.cpp

```diff
@@ -507,21 +507,44 @@ string CompilerStack::applyRemapping(string const& _path, string const& _context
 			return false;
 		return std::equal(_a.begin(), _a.end(), _b.begin());
 	};
+	// Try to find whether _a is a closer match for context _reference than _b
+	// Defaults to longest prefix in case of a tie.
+	auto isClosestContext = [](string const& _a, string const& _b, string const& _reference)
+	{
+		int a = _reference.compare(_a);
+		int b = _reference.compare(_b);
+		if (a == 0)
+			return true;
+		else if (b == 0)
+			return false;
+		else if (abs(a) == abs(b)) {
+			return a > 0;
+		}
+		return abs(a) < abs(b);
+	};
 
+	using filepath = boost::filesystem::path;
+	filepath context(_context);
 	size_t longestPrefix = 0;
 	string longestPrefixTarget;
+	string currentClosestContext;
+	string referenceContext = context.parent_path().generic_string();
 	for (auto const& redir: m_remappings)
 	{
+		filepath redirContext(redir.context);
 		// Skip if we already have a closer match.
-		if (longestPrefix > 0 && redir.prefix.length() <= longestPrefix)
+		if (longestPrefix > 0 && redir.prefix.length() < longestPrefix)
 			continue;
 		// Skip if redir.context is not a prefix of _context
-		if (!isPrefixOf(redir.context, _context))
+		if (!isPrefixOf(redirContext.generic_string(), _context))
 			continue;
 		// Skip if the prefix does not match.
 		if (!isPrefixOf(redir.prefix, _path))
 			continue;
-
+		// Skip if there is a prefix collision and the current context is closer
+		if (redir.prefix.length() == longestPrefix && !isClosestContext(redirContext.generic_string(), currentClosestContext, referenceContext))
```

Comment:
There is no need to store or compare the context as strings. In Line 519 / 539, we checked that `redir.context` is a prefix of `_context`, which means that the only real differences between contexts at this line here is their length. Unless I made a mistake, it suffices to store `redir.context.length()` and `isClosestContext` can be replaced by a simple integer comparison.